python/ray/dashboard/modules/node/node_head.py

```diff
@@ -170,104 +166,121 @@ def get_internal_states(self):
-                            key.encode(),
-                            namespace=ray_constants.KV_NAMESPACE_DASHBOARD,
-                            timeout=GCS_RPC_TIMEOUT_SECONDS,
-                        )
-                        if agent_port:
-                            agents[node_id] = json.loads(agent_port)
-                for node_id in agents.keys() - set(alive_node_ids):
-                    agents.pop(node_id, None)
-
-                DataSource.agents.reset(agents)
-                DataSource.nodes.reset(nodes)
+                published = await subscriber.poll(
+                    batch_size=node_consts.RAY_NODE_HEAD_SUBSCRIBER_POLL_SIZE
+                )
+                updated_dicts = await get_or_create_event_loop().run_in_executor(
+                    self._dashboard_head._thread_pool_executor,
+                    batch_updated_pairs_to_dict,
+                    published,
+                )
+                yield from updated_dicts
             except Exception:
-                logger.exception("Error updating nodes.")
-            finally:
-                self._node_update_cnt += 1
-                # _head_node_registration_time_s == None if head node is not
-                # registered.
-                head_node_not_registered = not self._head_node_registration_time_s
-                # Until the head node is registered, we update the
-                # node status more frequently.
-                # If the head node is not updated after 10 seconds, it just stops
-                # doing frequent update to avoid unexpected edge case.
+                logger.exception("Failed handling updated nodes.")
+
+    async def _update_node(self, node: dict):
+        node_id = node["nodeId"]  # hex
+        if node["isHeadNode"] and not self._head_node_registration_time_s:
+            self._head_node_registration_time_s = time.time() - self._module_start_time
+            # Put head node ID in the internal KV to be read by JobAgent.
+            # TODO(architkulkarni): Remove once State API exposes which
+            # node is the head node.
+            await self._gcs_aio_client.internal_kv_put(
+                ray_constants.KV_HEAD_NODE_ID_KEY,
+                node_id.encode(),
+                overwrite=True,
+                namespace=ray_constants.KV_NAMESPACE_JOB,
+                timeout=GCS_RPC_TIMEOUT_SECONDS,
+            )
+        assert node["state"] in ["ALIVE", "DEAD"]
+        is_alive = node["state"] == "ALIVE"
+        # Prepare agents for alive node, and pop agents for dead node.
+        if is_alive:
+            if node_id not in DataSource.agents:
+                # Agent port is read from internal KV. Problem is it's not present when
+                # we receive this update; it's only present after agent.py starts
+                # listening. So we make an async task that periodically polls internal
+                # KV.
```

Comment:
```suggestion
```suggestion
                # Agent port is read from internal KV, which is only populated 
                # upon Agent startup. In case this update received before agent
                # fully started up, we schedule a task to asynchronously update 
                # DataSource with appropriate agent port.
```
```